Grpc's event loops are kept separated from the Rest ones

### Description

While running a stress test against Quarkus Superheroes at https://github.com/quarkusio/quarkus-super-heroes/tree/main/grpc-locations I see few "towers" in the flamegraph which shows HTTP 2- related to GRPC handling  which are kept separated from the Netty event loops powering Quarkus's Rest. 
Since it's all non-blocking code I would expect to reuse the same event loops; this will save resources and simplify the configuration perks to users.



### Implementation ideas

_No response_

So locally when I remove that and run `./mvnw clean verify` the build fails:

```
[INFO] Results:
[INFO] 
[INFO] Tests run: 40, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jar:3.4.1:jar (default-jar) @ grpc-locations ---
[INFO] Building jar: /Users/edeandre/workspaces/quarkus/quarkus-super-heroes-main/grpc-locations/target/grpc-locations-1.0.jar
[INFO] 
[INFO] --- quarkus:3.18.2:build (default) @ grpc-locations ---
[INFO] [io.quarkus.kubernetes.deployment.KubernetesDeployer] Selecting target 'minikube' since it has the highest priority among the implicitly enabled deployment targets
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.881 s
[INFO] Finished at: 2025-02-07T14:52:05-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.18.2:build (default) on project grpc-locations: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.minikube.deployment.MinikubeProcessor#createConfigurators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.minikube.deployment.MinikubeProcessor.createConfigurators(MinikubeProcessor.java:90)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.KnativeProcessor#createConfigurators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.KnativeProcessor.createConfigurators(KnativeProcessor.java:130)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.VanillaKubernetesProcessor#createConfigurators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.VanillaKubernetesProcessor.createConfigurators(VanillaKubernetesProcessor.java:125)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.OpenshiftProcessor#createConfigurators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.OpenshiftProcessor.createConfigurators(OpenshiftProcessor.java:149)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.VanillaKubernetesProcessor#createDecorators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.getPort(KubernetesCommonHelper.java:163)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.getPort(KubernetesCommonHelper.java:156)
[ERROR]         at io.quarkus.kubernetes.deployment.VanillaKubernetesProcessor.createDecorators(VanillaKubernetesProcessor.java:184)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.minikube.deployment.MinikubeProcessor#createDecorators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.getPort(KubernetesCommonHelper.java:163)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.getPort(KubernetesCommonHelper.java:156)
[ERROR]         at io.quarkus.kubernetes.deployment.DevClusterHelper.createDecorators(DevClusterHelper.java:102)
[ERROR]         at io.quarkus.minikube.deployment.MinikubeProcessor.createDecorators(MinikubeProcessor.java:134)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)
[ERROR] 
[ERROR]         [error]: Build step io.quarkus.kubernetes.deployment.OpenshiftProcessor#createDecorators threw an exception: java.lang.IllegalArgumentException: All Kubernetes ports must be unique - 8089 has been used multiple times
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.verifyPorts(KubernetesCommonHelper.java:1169)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.combinePorts(KubernetesCommonHelper.java:181)
[ERROR]         at io.quarkus.kubernetes.deployment.KubernetesCommonHelper.getPort(KubernetesCommonHelper.java:163)
[ERROR]         at io.quarkus.kubernetes.deployment.OpenshiftProcessor.createDecorators(OpenshiftProcessor.java:228)
[ERROR]         at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:499)

```

--- a/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/config/GrpcClientConfiguration.java
+++ b/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/config/GrpcClientConfiguration.java
@@ -25,6 +25,12 @@ public interface GrpcClientConfiguration {
     boolean useQuarkusGrpcClient();
 
     /**
+     * Use Vert.x event loop(s) for gRPC client, if it's using the previous Java gRPC support.
+     */
+    @WithDefault("true")
+    boolean useVertxEventLoop();
+
+    /**
      * Configure XDS usage, if enabled.
      */
     @ConfigDocSection(generated = true)

--- a/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/Channels.java
+++ b/extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/supports/Channels.java
@@ -45,6 +45,7 @@ import io.grpc.MethodDescriptor;
 import io.grpc.netty.GrpcSslContexts;
 import io.grpc.netty.NegotiationType;
 import io.grpc.netty.NettyChannelBuilder;
+import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.handler.ssl.SslContext;
 import io.netty.handler.ssl.SslContextBuilder;
 import io.quarkus.arc.Arc;
@@ -206,8 +207,21 @@ public class Channels {
                 builder.defaultServiceConfig(map);
             }
 
-            if (builder instanceof NettyChannelBuilder) {
+            if (config.useVertxEventLoop() && builder instanceof NettyChannelBuilder) {
                 NettyChannelBuilder ncBuilder = (NettyChannelBuilder) builder;
+                // just use the existing Vertx event loop group, if possible
+                Vertx vertx = container.instance(Vertx.class).get();
+                // only support NIO for now, since Vertx::transport is not exposed in the API
+                if (vertx != null && vertx.isNativeTransportEnabled()) {
+                    // see https://github.com/eclipse-vertx/vert.x/pull/5292
+                    boolean reuseNettyAllocators = Boolean.getBoolean("vertx.reuseNettyAllocators");
+                    if (reuseNettyAllocators) {
+                        // let Netty Grpc to re-use the default Netty allocator as well
+                        System.setProperty("io.grpc.netty.useCustomAllocator", "false");
+                    }
+                    ncBuilder = ncBuilder.eventLoopGroup(vertx.nettyEventLoopGroup())
+                            .channelType(NioSocketChannel.class);
+                }
                 builder = ncBuilder
                         // clients are intercepted using the IOThreadClientInterceptor interceptor which will decide on which
                         // thread the messages should be processed.
@@ -393,6 +407,11 @@ public class Channels {
             @Override
             public boolean useQuarkusGrpcClient() {
                 return false;
+            }
+
+            @Override
+            public boolean useVertxEventLoop() {
+                return true;
             }
 
             @Override

